worker/uniter/relation/state.go

```diff
@@ -108,200 +107,196 @@ func (s *State) Validate(hi hook.Info) (err error) {
 	return nil
 }
 
-// StateDir is a filesystem-backed representation of the state of a
-// relation. Concurrent modifications to the underlying state directory
-// will have undefined consequences.
-type StateDir struct {
-	// path identifies the directory holding persistent state.
-	path string
-
-	// state is the cached state of the directory, which is guaranteed
-	// to be synchronized with the true state so long as no concurrent
-	// changes are made to the directory.
-	state State
+// UnitStateReadWriter encapsulates the methods from a state.Unit
+// required to set and get unit state.
+type UnitStateReadWriter interface {
+	State() (params.UnitStateResult, error)
+	SetState(unitState params.SetUnitStateArg) error
 }
 
-// State returns the current state of the relation.
-func (d *StateDir) State() *State {
-	return d.state.copy()
+// StateOps is a representation of the relation state. Concurrent
+// modifications to the underlying state will have undefined consequences.
+type StateOps struct {
```

Comment:
Is there a particular reason why we don't use a single object to load the complete state data at construction time (just like ReadAllStates) and then provide an API to read/write the data for a particular relation ID? This could be something created by the uniter and passed through to the relation state tracker and from there to the relationers.

This way, your data will always be up to date (and if the uniter restarts it will pull a fresh copy) and you don't need to have calls like `Ensure` in there.

Reply:
StateOps have now been replaced with StateManager which handles all of the relation state.